draft-irtf-cfrg-hpke.md

```diff
@@ -683,6 +686,25 @@ The general security properties of HPKE are described in
 {{security-properties}}.  In this section, we consider a security issue that may
 arise in practice and an advanced use case.
 
+## External Requirments / Non-Goals
+
+HPKE is designed to be a fairly low-level ompon, and thus does not provide
+several featues that a more high-level protocol might provide, for example:
+
+* Downgrade prevention - HPKE assumes that the initiator and receiver agree on
+  what algorithms to use.  Depending on how these algorithms are negotiated, it
+  may be possible for an intermediary to force the two parties to use
+  suboptimal algorithms.
+
+* Replay protection - The requirement that ciphertexts be presented to the
+  Context.Open funnction in the same order they were generated by Context.Seal
+  provides a degree of replay protection within a stream of ciphertexts
+  resulting from a given Context.  HPKE provides no other replay protection.
+
+* Forward secrecy - HPKE ciphertexts are not forward-secure.  A given ciphertext
+  can be decrypted if the recipient's public key is compromised at any time
+  after the ciphertext is created.
```

Comment:
The ciphertext also leaks if the public key is compromised before the ciphertext is created, so perhaps:

```
* Forward secrecy - HPKE ciphertexts are not forward-secure.  A given ciphertext
   can be decrypted if the recipient's public encryption key is compromised.
```

Reply:
In `mode_auth_psk`, the recipient's public key _and_ the psk need to be compromised. Suggestion:

```
* Forward secrecy - HPKE ciphertexts are not forward-secure. In mode_base
  and mode_auth, a given ciphertext can be decrypted if the recipient's public
  encryption key is compromised. In mode_psk and mode_auth_psk, a given
  ciphertext can be decrypted if the recipient's public encryption key and the
  psk are compromised.
```

Reply:
Good suggestion!

Reply:
Note that this confuses post-compromise security and forward secrecy.  See the recent discussion on the lake list on this point.  In short, a PSK mode could have FS, but cannot have PCS.

Reply:
Hmm, I don't think this is true. (I agree with the claim about achieving FS by hash-racheting the PSK.). @blipp's text simply says that if the private key material is compromised, bad things happen. It doesn't rule out FS should that keying material be updated. I think the text is fine as is.

Reply:
I am not on the lake list, so I don't know what was discussed there (just mentioning such that you can read the following with a grain of salt).

In `mode_psk` and `mode_auth_psk`, it's right that if _only one_ of (psk, skR) gets compromised after encrypting the plaintext, the plaintext stays secret. That's some weird kind of forward secrecy assuming that skR and psk are stored sufficiently separated (weird because new ciphertexts are still created with the same key material, there is no key update, no transition to a new epoch, etc). I have a proof for this property within my analysis. I think it's a good idea to mention this in the security considerations section, but maybe it's already implicit with my previous text suggestion.

A protocol building upon HPKE could try to add better properties, but I am not sure if this would be out of scope of this draft?